src/infer_request.h

```diff
@@ -679,10 +680,32 @@ class InferenceRequest {
     secondary_stats_aggregator_ = secondary_stats_aggregator;
   }
 
-  void Cancel() { response_factory_->Cancel(); }
-  void ResetCancel() { response_factory_->ResetCancel(); }
+  Status Cancel()
+  {
+    if (response_factory_) {
+      response_factory_->Cancel();
+    } else {
+      return Status(
+          Status::Code::INTERNAL,
+          "It is not possible to cancel an inference request before calling "
+          "TRITONSERVER_InferAsync.");
+    }
+    response_factory_->Cancel();
+    return Status::Success;
+  }
 
-  bool IsCancelled() { return response_factory_->IsCancelled(); }
+  Status IsCancelled(bool* is_cancelled)
+  {
+    if (response_factory_) {
```

Comment:
Question: in previous iteration - we would have had an issue if user called cancelled before setresponsecallback - is that correct?

Question: can user call inferasync without calling setresponse callback? does that error?

Reply:
>in previous iteration - we would have had an issue if user called cancelled before setresponsecallback - is that correct?

that's correct.

>Question: can user call inferasync without calling setresponse callback? does that error?

I think if it is not provided it would segfault when the backend tries to send responses: https://github.com/triton-inference-server/core/blob/a23786c25b36f07d2cfd510821682c117b50a99e/src/infer_response.cc#L219C11-L223

Reply:
> I think if it is not provided it would segfault when the backend tries to send responses: https://github.com/triton-inference-server/core/blob/a23786c25b36f07d2cfd510821682c117b50a99e/src/infer_response.cc#L219C11-L223

Maybe low hanging fruit for the future, but we could catch that error and return it for some improved UX instead of segfaulting for C-API and bindings users.

Reply:
`SetResponseCallback` is part of the request initialization, the server assumes the request is properly initialized when `InferAsync` is called.

I think we can draw the separation between before and after `InferAsync`, only after the call, infer request is passed into Triton realm. Before that, the user / frontend has all the control of infer request and calling `Cancel / IsCancelled` doesn't have real meaning.